src/ucp/core/ucp_am.c

```diff
@@ -1132,39 +1137,69 @@ ucp_am_handler_common(ucp_worker_h worker, ucp_am_hdr_t *am_hdr, size_t hdr_size
                       uint64_t recv_flags)
 {
     ucp_recv_desc_t *desc = NULL;
-    void *data;
+    ucp_am_entry_t *am_cb = &ucs_array_elem(&worker->am, am_hdr->am_id);
+    void *data            = ucp_am_data_ptr(am_hdr, hdr_size);
+    int need_desc;
     ucs_status_t status;
 
-    recv_flags |= (am_flags & UCT_CB_PARAM_FLAG_DESC) ?
+    ucs_assert(total_length >= am_hdr->header_length + hdr_size);
+
+    need_desc   = !(am_flags & UCT_CB_PARAM_FLAG_DESC) &&
+                  (am_cb->flags & UCP_AM_FLAG_PERSISTENT_DATA);
+
+    recv_flags |= ((am_flags & UCT_CB_PARAM_FLAG_DESC) || need_desc) ?
```

Comment:
i think it's possible to refactor and reduce number of branches in this function

Reply:
refactored